src/utils/mod.rs

```diff
@@ -647,29 +647,44 @@ impl TryFrom<Signature> for TPMT_SIGNATURE {
 }
 
 /// Rust native wrapper for session attributes objects.
+///
+/// If no mask has been specified then it will be
+/// defaulted to be have the same value as the flags.
 #[derive(Copy, Clone, Debug, Default)]
-pub struct TpmaSession(TPMA_SESSION);
+pub struct TpmaSession {
+    flags: TPMA_SESSION,
+    mask: Option<TPMA_SESSION>,
+}
 
 impl TpmaSession {
     /// Create a new session attributes object.
     pub fn new() -> TpmaSession {
-        TpmaSession(0)
+        TpmaSession {
+            flags: 0,
+            mask: None,
+        }
     }
 
     /// Set flag.
     pub fn with_flag(mut self, flag: TPMA_SESSION) -> Self {
-        self.0 |= flag;
+        self.flags |= flag;
+        self
+    }
+
+    /// Set a mask
+    pub fn with_mask(mut self, mask: TPMA_SESSION) -> Self {
```

Comment:
Do you think it'd be more useful to replace the existing mask completely and maybe have a separate method like `add_mask_flags` to add new ones? Don't know if there's any use case for removing flags.

Reply:
Hrrm... I think you are some what correct. Though the `with_mask` cannot clear the mask in  `TpmaSession` that would make behave differently then `with_flag` which would be confusing.

A `set_mask` would be logical for a function that clears the mask and sets it.

Reply:
Or rather I do not know. The `TpmaSession` struct is currently more of a builder. It would make more sense to split it up in a `TpmaSession` and a `TpmaSessionBuilder`. 

The `TpmaSession` should have the possibility to create a new `TpmaSession` out of an old one but with a specific mask.

Something like `copy_with_mask` or `mask_specific_copy` or something like that.


rust```
fn foo(tpma_session: &TpmaSession) {
context.tr_sess_set_attributes(session, tpma_session.copy_with_mask(MY_FIRST_MASK));
.
.
.
context.tr_sess_set_attributes(session, tpma_session.copy_with_mask(MY_SECOND_MASK));
}
```

I think that would make the most sense.

Reply:
I think the name was mostly drawn from the builder pattern, although in this case the structure is its own builder. Maybe it would make more sense to implement `From<TPMA_SESSION> for TpmaSession` and change these ones to just do the correct thing implied by their name (either overwrite or just set some particular flags) - i.e. not return `Self` as well

Reply:
Agree, would be ok with removing the "builder" methods and replacing them with setters and a `clone_with_mask(TPMA_SESSION)` method (clone and not copy because [copy is supposed to be implicit](https://doc.rust-lang.org/std/clone/trait.Clone.html))

Reply:
I do not really see a reason to remove the builder. I was thinking to have it like this:

```rust
#[derive(Copy, Clone, Debug, Default)]
pub struct TpmaSession {
    flags: TPMA_SESSION,
    mask: TPMA_SESSION,
}

impl TpmaSession {
    // Clones the TpmaSession but adds a new mask.
    pub fn clone_with_mask(self, mask: TPM_SESSION) -> TmaSession {
        TpmaSession {
            flags: self.flags,
            mask: mask,
        }
    }

    /// Get mask.
    pub fn mask(self) -> TPMA_SESSION {
        self.mask
    }

    /// Get all set flags.
    pub fn flags(self) -> TPMA_SESSION {
        self.flags
    }
}

/// A builder for TpmaSession.
///
/// If no mask have been added then the mask
/// will be set till all flags.
#[derive(Copy, Clone, Debug, Default)]
struct TpmaSessionBuilder {
    flags: TPMA_SESSION,
    mask: Option<TPMA_SESSION>,
}

impl TpmaSessionBuilder {
    pub fn new() -> TpmaSession {
        TpmaSession { flags: 0, mask: 0 }
    }

    /// Set flag.
    pub fn with_flag(mut self, flag: TPMA_SESSION) -> Self {
        self.flax |= flag;
        self
    }

    /// Set a mask
    pub fn with_mask(mut self, mask: TPMA_SESSION) -> Self {
        self.mask = (self.mask.unwrap_or(0) | mask);
        self
    }

    pub fn build(self) -> TpmaSession {
        TpmaSession {
            flags: self.flags,
            mask: self.mask.unwrap_or(self.flags),
        }
    }
}
```